### PR TITLE
ML&AI - LLM tutorials 1-4

### PR DESCRIPTION
**Pull request by WandrilleD**
@bebatut 

I have gone through the 4 tutorials:
 * https://training.galaxyproject.org/training-material//topics/statistics/tutorials/genomic-llm-pretraining/tutorial.html
 * https://training.galaxyproject.org/training-material//topics/statistics/tutorials/genomic-llm-finetuning/tutorial.html#evaluate-model-performance
 * https://training.galaxyproject.org/training-material//topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.html#compare-snps-effects-between-introns-and-exons
 * https://training.galaxyproject.org/training-material//topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.html#exploring-synthetic-gene-generation

and applied fixes, mainly to code, as well as added some solutions to questions where there were none.

### CODE REVIEW

File: topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md [hunk 1]
```diff
@@ -499,17 +503,27 @@ We will use the files `data_path` folder we just defined:
...
+from functions import *
+```
+
 We use the `SupervisedDataset` class to load and prepare the datasets. This class handles the tokenization and formatting of the data, making it ready for model training and evaluation.
 
 ```python
 train_dataset = SupervisedDataset(
     tokenizer=tokenizer,
-    data_path=Path(data_path / "train.csv"),
+    data_path=os.path.join(data_path , "train.csv"),
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md:521]: I would prefer to keep `pathlib` as it is more of a best practice. It is much easier and cleaner to work with, especially for cross-platform development.

File: topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md [hunk 2]
```diff
@@ -591,11 +605,10 @@ model.config.pad_token_id = tokenizer.pad_token_id
 
 We can now set up the `Trainer` to manage the training and evaluation process of our model. The `Trainer` class simplifies the training loop, handling many of the complexities involved in training deep learning models.
 
-Before setting up the `Trainer`, we load custom function (`compute_metrics`) to compute metrics for `Trainer` stored in the a `scriptPython/functions.py`:
+We first need to attach the LoRA adapter to the model:
 
 ```python
-sys.path.append("scriptPython/")
-from functions import *
+model.add_adapter(peft_config , adapter_name="lora_1" )
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md:611]:
```suggestion
model.add_adapter(peft_config, adapter_name="lora_1")
```

File: topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md [hunk 3]
```diff
@@ -702,17 +715,6 @@ eval_steps_per_second          9.620000
 epoch                          3.000000
 ```
 
-> <question-title></question-title>
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md:705]: I forgot to do it. But the idea was to move the above output and comment it in a question format

File: topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md [hunk 4]
```diff
@@ -740,7 +742,7 @@ For that, 80% of the entire data will be used for the training set and the remai
 
 ```python
 train_size = int(0.8 * len(dataset["train"]))
-val_size = len(dataset["train"]) - training_size
+val_size = len(dataset["train"]) - train
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md:745]:
```suggestion
val_size = len(dataset["train"]) - train_size
```

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 5]
```diff
@@ -358,25 +358,16 @@ We now compute the maximum of the hidden states accross the sequence length dime
 embedding_max = torch.max(hidden_states, dim=1)[0]
 ```
 
-> <question-title></question-title>
->
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:362]: It might be good to use this question box to check the value and comment it

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 6]
```diff
@@ -445,27 +437,30 @@ def downloadReadFastaFile(fasta_file):
       print(f"Failed to download file. HTTP Status code: {response.status_code}")
   # Read the file
   seql_list=[]
-  with gzip.open(fasta_file, "rt") as handle:
+  with gzip.open('file.gz', "rt") as handle:
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:440]:
```suggestion
  with gzip.open("file.gz", "rt") as handle:
```

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 7]
```diff
@@ -515,7 +516,7 @@ To compute the effect of SNPs, we need :
   def computeMutationEffect(wt_seqs, mut_seqs):
     wt_embedding = computeEmbedding(wt_seqs)
     mut_embedding = computeEmbedding(mut_seqs)
-    return torch.norm(mut_embedding - wt_embedding)
+    return torch.norm(mut_embedding - wt_embedding , dim=1)
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:519]:
```suggestion
    return torch.norm(mut_embedding - wt_embedding, dim=1)
```

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 8]
```diff
@@ -530,7 +531,8 @@ intron_SNP_distL2 = computeMutationEffect(intron_wt_seqs,intron_mut_seqs)
 >
 > > <solution-title></solution-title>
 > >
-> >  
+> > 10 000 each : one value per pair of wt/mutated sequence
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:534]:
```suggestion
> > 10,000 each: one value per pair of wt/mutated sequence
```

**Review by bebatut — COMMENTED** [review]
Thanks @WandrilleD. I made some small suggestions

File: topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md (hunk 1 above)
Comment (WandrilleD) [reply topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md:521]: OK, although `os.path` is cross-platform too. 
Anyhow, the most important is then to use the correct syntax for `pathlib`:
`data_path=Path(data_path ) / "train.csv" ,`

**Review by bebatut — APPROVED** [review]
Thanks @WandrilleD for these changes!
